### PR TITLE
implemented Selafin class from selafin.py

### PR DESCRIPTION
**Pull request by tomsail**
works for 2D and 3D
tests are working
Lazy index yet to be implemented correctly with (or without) dask

### CODE REVIEW

File: xarray_selafin_backend/xarray_backend.py [hunk 1]
```diff
@@ -32,23 +42,27 @@ def _raw_indexing_method(self, key):
...
-        print(key)
 
         # Convert time_key and node_key to ranges to handle steps and to list indices for SELAFIN reader
-        time_indices = range(*time_key.indices(self.shape[0])) if isinstance(time_key, slice) else [time_key]
-        node_indices = range(*node_key.indices(self.shape[1])) if isinstance(node_key, slice) else [node_key]
+        time_indices = (
+            range(*time_key.indices(self.shape[0]))
+            if isinstance(time_key, slice)
+            else [time_key]
+        )
```
Comment (pmav99) [inline xarray_selafin_backend/xarray_backend.py:51]: A ternary operator that actually needs one extra line compared to a normal if/else is a prime example of "when not to use ternary operators!":
``` python
if isinstance(time_key, slice):
    time_indices = range(*time_key.indices(self.shape[0]))
else:
    time_indices = [time_key]
```
Furthermore, I recommend avoiding using ternary operators in code that you plan to test. 

The issue with ternary operators is that they merge two execution paths into one line, leading to potentially deceptive code coverage. It’s possible that your tests may only examine one of the paths, leaving you unaware of the untested path unless you specifically notice it

File: xarray_selafin_backend/xarray_backend.py [hunk 2]
```diff
@@ -32,23 +42,27 @@ def _raw_indexing_method(self, key):
...
+        node_indices = (
+            range(*node_key.indices(self.shape[1]))
+            if isinstance(node_key, slice)
+            else [node_key]
+        )
```
Comment (pmav99) [inline xarray_selafin_backend/xarray_backend.py:56]: ditto (ternary operators)

File: xarray_selafin_backend/xarray_backend.py (hunk 1 above)
Comment (tomsail) [reply xarray_selafin_backend/xarray_backend.py:51]: indeed. 
the problem I have before even treating the case of each key is to determine how to treat the keys themselves. 
```python
        if isinstance(key, tuple):
            time_key, node_key, plan_key = key
        else:
            # Assuming if only one key is provided, it's for the time dimension
            time_key = key
            # And we want to select all nodes
            node_key = slice(None)
            # and all plans 
            plan_key = slice(None)
``` 
this feels insufficient if I want to slice only with time or only with the nodes index, or finally only with the plans (in 3D).

Comment (tomsail) [reply xarray_selafin_backend/xarray_backend.py:51]: isn't there any way to recognize which is the key used?

Comment (tomsail) [reply xarray_selafin_backend/xarray_backend.py:51]: isn't there any way to recognize which is the key used?

**Comment by pmav99** [comment]
After the latest commits all the tests are broken. 
```
FAILED tests/io_test.py::test_open_dataset[3D] - NotImplementedError: Lazy array not yet implemented
FAILED tests/io_test.py::test_open_dataset[2D] - NotImplementedError: Lazy array not yet implemented
FAILED tests/io_test.py::test_to_netcdf[3D] - NotImplementedError: Lazy array not yet implemented
FAILED tests/io_test.py::test_to_netcdf[2D] - NotImplementedError: Lazy array not yet implemented
FAILED tests/io_test.py::test_to_selafin[3D] - NotImplementedError: Lazy array not yet implemented
FAILED tests/io_test.py::test_to_selafin[2D] - NotImplementedError: Lazy array not yet implemented
```
Did you forget to commit something?

**Comment by tomsail** [comment]
great.. well dask shouldn't be implemented (it is now optional). So I might have forgotten to update pyproject.toml

I'll fix this and add LazyIndex support (I almost got it working now)

**Comment by tomsail** [comment]
not dask + lazy index works but the precision from writing respectively in `slf`/ `netcdf` changes 

see the function I implemented in io_test.py

```
Values for variable 'VELOCITY U' differ.
Values for variable 'VELOCITY V' differ.
Values for variable 'WATER DEPTH' differ.
Values for variable 'FREE SURFACE' differ.
Values for variable 'BOTTOM' differ.
```

**Pull request closed without merging**